### ISSUE TITLE
Add global shuffle

### ISSUE THREAD

**Issue by sagiahrac** [Eventual-Inc/Daft#2612]
**Is your feature request related to a problem? Please describe.**
Daft does not provide a native operation for global shuffle. The `.sample` method doesn't change the rows order, and `.repartition` keeps the original order of rows with respect to the original dataframe in each partition.
```python
df_a = daft.from_pydict({"char": [f"a{i}" for i in range(10)]})
df_b = daft.from_pydict({"char": [f"b{i}" for i in range(10)]})

df = df_a.concat(df_b)
df.repartition(1).to_pydict()
```

```
{'char': ['a0',
  'a1',
  'a2',
  'a3',
  'a4',
  'a5',
  'a6',
  'a7',
  'a8',
  'a9',
  'b0',
  'b1',
  'b2',
  'b3',
  'b4',
  'b5',
  'b6',
  'b7',
  'b8',
  'b9']}
```

**Describe the solution you'd like**
After assigning each row a random partition, shuffle each partition. Add an optional way to control the seed.

**Describe alternatives you've considered**
1. Add a column of random integers and sort the dataframe by it (requires calculating the size of that dataframe)
2. Materializing the dataframe as shuffle is a global operation anyways and shuffle with polars (with `.sample`)
3. Iterating over partitions and shuffling each partition: 

> ```python
> class MyDataset(torch.utils.data.IterableDataset):
>     df: daft.DataFrame
>     ...
> 
>     def __iter__(self):
>         df = self.df.repartition(None)
>         for partition in df.iter_partitions():
>             samples = partition.to_pylist()
>             np.random.shuffle(samples)
>             yield from samples
> ```

**Comment by ccmao1130** [comment]
Slack thread on Daft for more context: https://dist-data.slack.com/archives/C041NA2RBFD/p1720707378027859

**Comment by jonasdedden** [comment]
@colin-ho is there any ETA on this feature?

**Comment by srilman** [comment]
Hi @jonded94, to clarify, 
* Are you looking for a completely randomly row rearrangement / shuffle, similar to https://huggingface.co/docs/datasets/process#shuffle? That might be computationally expensive, because it would require materializing all rows in memory and doing the rearrangement
* Alternatively, we could support something similar to a https://huggingface.co/docs/datasets/v4.8.4/en/package_reference/main_classes#datasets.IterableDataset.shuffle, which requires a temporary buffer to sample and reorder rows

**Comment by jonasdedden** [comment]
Hey @srilman, thank you for reaching out to me! Actually, both would be interesting for us. 

The former would be definitely helpful, as there are tasks where you want to *globally* and totally randomly shuffle a big (>TiBs) dataset, and existing tooling besides Daft(-on-Ray) sometimes behaves unexpectedly poorly, even with clusters with TiBs of RAM. Any new solution to try out (spilling data to disk for larger than memory workloads also would be super fine) would be interesting for us. Currently, we drive some custom workarounds with manual sharding / hash bucketing, local shuffles and then merging of intermediate artifacts in the end.

The latter probably would more helpful for introspection / exploration of data, but also sounds nice.

**Comment by srilman** [comment]
Sounds good. I don't think it should be too hard to support the former, as we can simulate it by generating a random number per row followed by a sort by the random number. We already have pretty good sorts, so adding a random value generator on top shouldn't be too bad.

From there, we can extend with the latter option, and maybe even a mix of the 2 at some point. That seems a bit harder, so maybe we can do it as a next step.

Let me see what I can whip up.

**Comment by desmondcheongzx** [comment]
Thanks for the feature request, @sagearc! This has been implemented by @srilman -- Daft now has a native `df.shuffle(seed=None)` API that globally reorders rows, with optional seed for reproducibility.

We also have a number of ideas internally around more efficient shuffle strategies for dataloading, but we'll track those in dedicated issues. Closing this out!
